R/age_model_class.R

```diff
@@ -0,0 +1,222 @@
+#'     model.
+#' @slot parameter_names names of the initial conditions and parameters
+#'     which are used by the model.
+#' @slot parameters named list containing parameters of the model.
+#'     Initial values for each compartment, S0, E0, I0, R0.
+#'     Other parameters: a, b, c which represent the rates of changes
+#'     between the compartments.
+#' @slot n_age_categories number of age categories.
+#'
+#' @import deSolve
+#' @import glue
+#' @import ggplot2
+
+setClass('age_model',
+         #slots
+         slots = c(
+           name = 'character',
+           output_names = 'list',
+           parameter_names = 'list',
+           parameters = 'list',
+           n_age_categories = 'numeric'
+         ),
+
+         #prototypes for the slots, automatically set output and param
+         #names
+         prototype = list(
+           name = NA_character_,
+           output_names = list('S', 'E', 'I', 'R', 'Incidence'),
+           parameter_names = list('S0', 'E0', 'I0', 'R0', 'a', 'b', 'c'),
+           parameters = vector(mode = "list", length = 7),
+           n_age_categories = NA_real_
+
+         )
+)
+
+#' @describeIn Retrieves parameters for an age-structured simple SEIR model.
+#'
+#' @param object An object of the class age_model.
+setGeneric('get_parameters', function(object) standardGeneric('get_parameters'))
+
+setMethod(
+  'get_parameters', 'age_model',
+  function(object) object@parameters
+)
+
+#' @describeIn Sets parameters for an age-structured simple SEIR model.
+#'
+#' Default parameters a = 1, b = 1 and c = 1.
+#' If the initial conditions provided to do not sum to 1 or of different
+#' sizes compared to the number of age groups, an error is thrown.
+#'
+#' @param object An object of the class age_model.
+#' @param S0 initial fraction of the population that is susceptible
+#'           by age group.
+#' @param E0 initial fraction of the population that has been exposed
+#'           by age group.
+#' @param I0 initial fraction of the population that is infected
+#'           by age group.
+#' @param R0 initial fraction of the population that has recovered
+#'           by age group.
+#'
+#' All initial conditions must sum up to 1.
+#'
+#' @param a rate at which an infected individual exposes susceptible
+#' @param b rate at which exposed individuals become infected
+#' @param c rate at which infected individuals recover
+#'
+#' All rates of change between compartments are equal regardless of
+#' age group.
+#'
+#' @return updated version of the age-structured SEIR model.
+setGeneric(
+  'set_parameters',
+  function(object, S0, E0, I0, R0, a = 1, b = 1, c = 1){
+    standardGeneric('set_parameters')
+  })
+
+setMethod(
+  'set_parameters', 'age_model',
+  function(object, S0, E0, I0, R0, a, b, c) {
+
+    # check that ICs are valid
+    if (sum(S0, E0, I0, R0) != 1) {
+      stop('Invalid initial conditions. Must add up to 1.')
+    }
+
+    #create list of parameter values
```

Comment:
A minor thing, but I'd try to put spaces between the hash and the start of the comment as it makes readability easier.

Reply:
This has been implemented throughout

Reply:
Thanks!